api/host_volumes.go

```diff
@@ -0,0 +1,236 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package api
+
+import "net/url"
+
+// HostVolume represents a Dynamic Host Volume: a volume associated with a
+// specific Nomad client agent but created via API.
+type HostVolume struct {
+	// Namespace is the Nomad namespace for the host volume, which constrains
+	// which jobs can mount it.
+	Namespace string
+
+	// ID is a UUID-like string generated by the server.
+	ID string
+
+	// Name is the name that group.volume will use to identify the volume
+	// source. Not expected to be unique.
+	Name string
+
+	// PluginID is the name of the host volume plugin on the client that will be
+	// used for creating the volume. If omitted, the client will use its default
+	// built-in plugin.
+	PluginID string `mapstructure:"plugin_id" hcl:"plugin_id"`
+
+	// NodePool is the node pool of the node where the volume is placed. If the
+	// user doesn't provide a node ID, a node will be selected using the
+	// NodePool and Constraints. If the user provides both NodePool and NodeID,
+	// NodePool will be used to validate the request. If omitted, the server
+	// will populate this value in before writing the volume to Raft.
+	NodePool string `mapstructure:"node_pool" hcl:"node_pool"`
+
+	// NodeID is the node where the volume is placed. If the user doesn't
+	// provide a NodeID, one will be selected using the NodePool and
+	// Constraints. If omitted, this field will then be populated by the server
+	// before writing the volume to Raft.
+	NodeID string `mapstructure:"node_id" hcl:"node_id"`
+
+	// Constraints are optional. If the NodeID is not provided, the NodePool and
+	// Constraints are used to select a node. If the NodeID is provided,
+	// Constraints are used to validate that the node meets those constraints at
+	// the time of volume creation.
+	Constraints []*Constraint `json:",omitempty" hcl:"constraint"`
+
+	// Because storage may allow only specific intervals of size, we accept a
+	// min and max and return the actual capacity when the volume is created or
+	// updated on the client
+	RequestedCapacityMin int64 `mapstructure:"capacity_min" hcl:"capacity_min"` // bytes
+	RequestedCapacityMax int64 `mapstructure:"capacity_max" hcl:"capacity_max"` // bytes
+	Capacity             int64 // bytes
```

Comment:
We're not really constrained here by CSI. And because of the quirk that we only use the `api` package for serializing into the request body and not the response, we should keep this consistent between structs/api for the same type.